### PR TITLE
feat: InstanceTracker for FrameworkTemplatePool

### PR DESCRIPTION
**Pull request by ebariche**
## PR Type

What kind of change does this PR introduce?
- Feature

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 9a3c4e3</samp>

Improved memory management and performance for XAML templates on Skia by using weak references and moving recycling logic to `FrameworkTemplatePool` and `InstanceTracker`. Removed finalizer and `Cleanup` method from `UIElement`.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs [hunk 1]
```diff
@@ -437,5 +487,60 @@ public TemplateEntry(TimeSpan creationTime, View dependencyObject)
 
 			public View Control { get; private set; }
 		}
+
+		private static class InstanceTracker
+		{
+			private static ConditionalWeakTable<object /* parent */, TrackerCookie> _instances = new();
+
+			private static Stack<TrackerCookie> _cookiePool = new();
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs:495]: ConcurrentStack here is as well?

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs [hunk 2]
```diff
@@ -437,5 +487,60 @@ public TemplateEntry(TimeSpan creationTime, View dependencyObject)
...
+					_instance = instance;
+				}
+
+				~TrackerCookie()
+				{
+					Instance.RaiseOnParentCollected(_instance!);
+
+					_instance = null;
+
+					GC.ReRegisterForFinalize(this);
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs:537]: This is somewhat confusing to me.

File: src/Uno.UI/UI/Xaml/UIElement.skia.cs [hunk 3]
```diff
@@ -46,11 +46,6 @@ public UIElement()
 			UpdateHitTest();
 		}
 
-		~UIElement()
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/UIElement.skia.cs:49]: `UIElement.wasm.cs` has something similar. I think the new code will be applicable to Wasm and we can remove the UIElement finalizer on Wasm as well?

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs (hunk 2 above)
Comment (ebariche) [reply src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs:579]: Correct !

File: src/Uno.UI/UI/Xaml/UIElement.skia.cs (hunk 3 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/UIElement.skia.cs:49]: Wasm needs to keep its finalizer for now, as the purpose is different (clearing native resources).

**Review by jeromelaban — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs [hunk 4]
```diff
@@ -437,5 +488,101 @@ public TemplateEntry(TimeSpan creationTime, View dependencyObject)
...
+
+					_handles.Add(new DependentHandle(parent, cookie));
+				}
+			}
+
+			public static void ReturnCookie(TrackerCookie cookie)
+			{
+				lock (_cookiePool)
+				{
+					_cookiePool.Push(cookie);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs:560]: Let's make sure to trim excess in _cookiePool.

**Review by MartinZikmund — APPROVED** [review]

**Pull request merged**